### ISSUE TITLE
Loss minimization with dice_coe

### ISSUE THREAD

**Issue by jackyko1991** [tensorlayer/tensorlayer#891]
### New Issue Checklist

- [x] I have read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I searched for [existing GitHub issues](https://github.com/tensorlayer/tensorlayer/issues)

### Issue Description

I am trying to use [dice_coe] (https://tensorlayer.readthedocs.io/en/stable/_modules/tensorlayer/cost.html) as a loss function after the argmax of network logit output. However when I implement optimization on the loss, the argmax seems not a differentiable function and unable to perform gradient descent.

### Reproducible Code

```python

# Initialize the model
        with tf.name_scope("vnet"):
            logits = VNet.v_net(images_placeholder,input_channels = input_batch_shape[4], output_channels =2)

# Dice Similarity
with tf.name_scope("dice"):
            sorensen = dice_coe(tf.expand_dims(pred,-1),labels_placeholder, loss_type='sorensen')
            jaccard = dice_coe(tf.expand_dims(pred,-1),labels_placeholder, loss_type='jaccard')
            sorensen_loss = tf.subtract(tf.constant(1.),sorensen) 
            jaccard_loss = tf.subtract(tf.constant(1.),jaccard) 
        tf.summary.scalar('sorensen', sorensen)
        tf.summary.scalar('jaccard', jaccard)
        tf.summary.scalar('sorensen_loss', sorensen_loss)
        tf.summary.scalar('jaccard_loss',jaccard_loss)

# Argmax Op to generate label from logits
        with tf.name_scope("predicted_label"):
            pred = tf.argmax(logits, axis=4 , name="prediction")

optimizer = tf.train.GradientDescentOptimizer(learning_rate=FLAGS.init_learning_rate)
            train_op = optimizer.minimize(
                loss=sorensen_loss,
                global_step=global_step)
# training cycle
with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  print("{}: Start training...".format(datetime.datetime.now()))

  train, summary = sess.run([train_op, summary_op], feed_dict={images_placeholder: image, labels_placeholder: label})
  train_summary_writer.add_summary(summary, global_step=tf.train.global_step(sess, global_step))


# ======================================================== #
###### THIS CODE IS AN EXAMPLE, REPLACE WITH YOUR OWN ######
# ======================================================== #
```

This structure works if i use tf original loss functions:
```Python
with tf.name_scope("cross_entropy"):
            loss_op = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
                logits=logits,
                labels=tf.squeeze(labels_placeholder, 
                squeeze_dims=[4])))
        tf.summary.scalar('loss',loss_op)
```

Any way I can use the argmax result to calculate dice loss then perform GD on that?

**Comment by DEKHTIARJonathan** [comment]
Hi @jackyko1991. Thanks for your bug report.

Unfortunately, this code miss some part (e.g. VNet) and has a lot of parts non necessary.

Can you please provide a **minimal and complete** code with the expected result, that we can try to understand the issue

Thanks

**Comment by zsdonghao** [comment]
feel free to reopen if you want to discuss.

we just release 2.0.0 version
